api/services/CasebookService.js

```diff
@@ -1,52 +1,93 @@
-const request = require('request');
-const requestPromise = require('request-promise');
-const sails = require('sails');
+const axios = require('axios');
+const https = require('https');
+
+const queryParamObjToStr = require('../helpers/queryParamObjToStr');
+const config = require('../../config/environment-variables');
+
+const {
+    hmacKey,
+    casebookCertificate: cert,
+    casebookKey: key,
+    customURLs,
+} = config;
 
 const CasebookService = {
-    get(options) {
-        return CasebookService._returnRequestMethods('get', options);
-    },
+    getApplicationStatus(applicationReference) {
+        const queryParamsObj = {
+            timestamp: Date.now().toString(),
+            applicationReference,
+        };
+        const casebookRequestBase =
+            CasebookService._createBaseRequest(queryParamsObj);
 
-    post(options) {
-        return CasebookService._returnRequestMethods('post', options);
+        return casebookRequestBase.get(customURLs.applicationStatusAPIURL, {
+            params: queryParamsObj,
+        });
     },
 
-    _returnRequestMethods(method, options) {
-        const requestType = options.promise ? requestPromise : request;
-        delete options.promise;
-
-        const authParams = CasebookService._createAuthParams(options.qs);
-        const optionsWithAuthParams = { ...options, ...authParams };
+    getApostilleDownload(apostilleReference) {
+        const queryParamsObj = {
+            timestamp: Date.now().toString(),
+            apostilleReference,
+        };
+        const casebookRequestBase =
+            CasebookService._createBaseRequest(queryParamsObj);
 
-        return requestType[method](optionsWithAuthParams);
+        return casebookRequestBase.get(customURLs.apostilleDownloadAPIURL, {
+            params: queryParamsObj,
+            responseType: 'stream',
+        });
     },
 
-    _createAuthParams(queryParamsObj) {
-        const {
-            hmacKey,
-            casebookCertificate: cert,
-            casebookKey: key,
-        } = sails.config;
+    _createBaseRequest(paramsObj) {
+        const casebookRequestBase = axios.create({
+            baseURL: customURLs.casebookBaseUrl,
+            headers: {
+                'Content-Type': 'application/json; charset=utf-8',
+                'api-version': '4',
+                hash: CasebookService._addHashToHeader(paramsObj),
+            },
+            httpsAgent: new https.Agent({
+                cert,
+                key,
+                keepAlive: true,
+            }),
+            paramsSerializer: queryParamObjToStr,
+            transformRequest: [CasebookService._addHmacToQueryParam],
+        });
+
+        casebookRequestBase.interceptors.response.use(
+            (response) => {
+                if (response.status !== 200) {
+                    throw new Error(`Casebook returned ${response.statusCode}`);
+                }
+                return response;
```

Comment:
this means that any other status code, even if it's 2xx (so for example, 204) will result in an error. Is that right? 

this is how axios can resolve ranges 
```
axios.get('/user/12345', {
  validateStatus: function (status) {
    return status < 500; // Resolve only if the status code is less than 500
  }
})
```

Reply:
Nah this is for sure an issue. I like the look of `validateStatus` but the default function seems to be exactly what I need:
```
  validateStatus: function (status) {
    return status >= 200 && status < 300; // default
  },
```

Maybe I should just get rid of my implementation 🤷

Reply:
the less code you have to write, the better ;)